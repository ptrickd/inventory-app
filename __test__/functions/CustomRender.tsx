//Testing library
import { render, RenderOptions } from "@testing-library/react";

//Mocked Files
import { mockedRouter } from "./mockedRouter";
// import { RouterContext } from "next/dist/shared/lib/router-context";
import { RouterContext } from "next/dist/shared/lib/router-context.shared-runtime";
//Context
import { UserContext } from "../../contexts/UserContext";
import { ReportsContext } from "../../contexts/ReportsContext";
import { CategoriesContext } from "../../contexts/CategoriesContext";
import { ProductsContext } from "../../contexts/ProductsContext";

//ContextProps
import {
  CategoriesContextProps,
  ProductsContextProps,
  UserContextProps,
  ReportsContextProps,
} from "./mockedContexts";

const AllTheProviders: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  return (
    <RouterContext.Provider value={mockedRouter({})}>
      <UserContext.Provider value={UserContextProps}>
        <ReportsContext.Provider value={ReportsContextProps}>
          <CategoriesContext.Provider value={CategoriesContextProps}>
            <ProductsContext.Provider value={ProductsContextProps}>
              {children}
            </ProductsContext.Provider>
          </CategoriesContext.Provider>
        </ReportsContext.Provider>
      </UserContext.Provider>
    </RouterContext.Provider>
  );
};

const CustomRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export default CustomRender;
