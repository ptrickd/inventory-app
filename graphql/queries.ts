//GraphQL
import { gql } from "@apollo/client";

//Authentication
export const REGISTER = gql`
  mutation Register($email: String!, $password: String!) {
    register(email: $email, password: $password) {
      user {
        id
        email
        theme
      }
      error
    }
  }
`;
export const LOGIN = gql`
  mutation Login($email: String!, $password: String!) {
    login(email: $email, password: $password) {
      token
      user {
        id
        email
        theme
      }
      error
    }
  }
`;

export const GET_CURRENT_USER = gql`
  query CurrentUser {
    currentUser {
      id
      email
      theme
    }
  }
`;

//Product
export const GET_PRODUCTS_BY_CATEGORY = gql`
  query ProductsByCategory($categoryId: String) {
    productsByCategory(categoryId: $categoryId) {
      id
      name
      currentAmount
      previousAmount
      categoryId
      unit
    }
  }
`;
export const CREATE_PRODUCT = gql`
  mutation CreateProduct($name: String!, $categoryId: String!, $unit: String!) {
    createProduct(name: $name, categoryId: $categoryId, unit: $unit) {
      id
      name
      categories {
        categoryId
        currentAmount
        previousAmount
      }
      unit
      error
    }
  }
`;
export const DELETE_PRODUCT = gql`
  mutation DeleteProduct($productId: ID!) {
    deleteProduct(productId: $productId) {
      id
      name
    }
  }
`;
export const EDIT_PRODUCT = gql`
  mutation EditProduct(
    $productId: ID
    $name: String
    $categoryId: String
    $unit: String
  ) {
    editProduct(
      productId: $productId
      name: $name
      categoryId: $categoryId
      unit: $unit
    ) {
      id
      name
      categoryId
      unit
    }
  }
`;

//Category
export const GET_CATEGORIES = gql`
  query GetCategories {
    categories {
      id
      name
      error
    }
  }
`;

export const GET_NUM_CATEGORIES = gql`
  query GetNumOfCategories {
    numOfCategories
  }
`;

export const CREATE_CATEGORY = gql`
  mutation CreateCategory($name: String!) {
    createCategory(name: $name) {
      id
      name
      error
    }
  }
`;

export const GET_CATEGORY = gql`
  query GetCategory($categoryId: ID!) {
    category(categoryId: $categoryId) {
      id
      name
    }
  }
`;
