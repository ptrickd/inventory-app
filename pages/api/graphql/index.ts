
//GraphQL
import { ApolloServer, gql } from 'apollo-server-micro'

//Controller
import {
    createProduct,
    getProducts,
    getProductsByCategory,
    editProduct,
    deleteProduct
} from '../../../controllers/product.controller'
import {
    getCategories,
    createCategory,
    getCategory,
    editCategory,
    deleteCategory
} from '../../../controllers/category.controller'

const typeDefs = gql`
    type Product {
        id: ID
        name: String
        amount: Int
        categoryId: ID
    }

    type Category {
        id: ID
        name: String
    }

    type Query {
        getProducts: [Product]
        getProductsByCategory(categoryId: String): [Product]

        getCategory(categoryId: ID):Category
        getCategories: [Category]
    }

    type Mutation {
        createProduct(name:String, amount:Int, categoryId: String): Product
        editProduct(id:ID, name:String, categoryId: String): Product
        deleteProduct(productId: ID): Product
        
        createCategory(name:String): Category
        editCategory(categoryId:ID, name:String): Category
        deleteCategory(categoryId:ID): Category
    }
`

interface IProduct {
    id: string
    name: string
    amount: number
    categoryId: string
}

interface ICreateProduct {
    name: string
    amount: number
    categoryId: string
}

interface IIds {
    [propName: string]: string
}

interface IEditProduct {
    id: string
    name: string
    categoryId: string
}

interface ICategory {
    categoryId: string
    name: string
}


const resolvers = {
    Query: {
        getProducts: async () => {
            try {
                const products = await getProducts()
                return products.map(({ id, amount, name, categoryId }: IProduct) => ({
                    id,
                    name,
                    amount,
                    categoryId
                }))
            } catch (err) {
                console.log(err)
            }
        },
        getProductsByCategory: async (_: any, { categoryId }: IIds) => {
            try {
                const products = await getProductsByCategory(categoryId)
                if (!products) throw new Error('No products found')
                return products.map(({ id, amount, name, categoryId }: IProduct) => ({
                    id,
                    name,
                    amount,
                    categoryId
                }))
            }
            catch (err) {
                console.log(err)
            }
        },
        getCategory: async (_: any, { categoryId }: IIds) => {
            try {
                const category = await getCategory(categoryId)
                if (!category) throw new Error("No Category Found");

                return category
            }
            catch (err) {
                console.log(err)
            }
        },
        getCategories: async () => {
            try {
                console.log('in getCategories')
                const categories = await getCategories()
                if (!categories) throw new Error("No Categories Found")
                return categories.map(({ id, name }) => ({
                    id, name
                }))
            }
            catch (err) {
                console.log(err)
            }
        }
    },
    Mutation: {
        createProduct: async (_: any, { name, amount, categoryId }: ICreateProduct) => {

            let product = await createProduct(name, amount, categoryId)
            return product
        },
        editProduct: async (_: any, { id, name, categoryId }: IEditProduct) => {
            let product = await editProduct(id, name, categoryId)
            if (!product) throw new Error('No product found')
            return product
        },
        deleteProduct: async (_: any, { productId }: IIds) => {
            let product = await deleteProduct(productId)
            if (!product) throw new Error('No product found')
            return product
        },
        createCategory: async (_: any, { name }: IIds) => {
            console.log('createCategory', name)
            let category = await createCategory(name)
            if (!category) throw new Error("No Category Create")
            return category

        },
        editCategory: async (_: any, { categoryId, name }: ICategory) => {
            let editedCategory = await editCategory(categoryId, name)
            if (!editedCategory) throw new Error("No Category Found")
            return editedCategory
        },
        deleteCategory: async (_: any, { categoryId }: IIds) => {
            let deletedCategory = await deleteCategory(categoryId)
            if (!deletedCategory) throw new Error("No Category Found")
            return deletedCategory
        }
    }
}


const apolloServer = new ApolloServer({ typeDefs, resolvers })

export const config = {
    api: {
        bodyParser: false
    }
}

export default apolloServer.createHandler({ path: "/api/graphql" })
