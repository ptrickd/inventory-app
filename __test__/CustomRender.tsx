//Testing library
import { render, RenderOptions } from "@testing-library/react";

//Mocked Router
import { MockedProvider } from "@apollo/client/testing";
import { mockedRouter } from "./mockedRouter";
import { RouterContext } from "next/dist/shared/lib/router-context";

//Context
import { UserContext } from "../contexts/UserContext";
import { ReportsContext } from "../contexts/ReportsContext";
import { CategoriesContext } from "../contexts/CategoriesContext";
import { ProductsContext } from "../contexts/ProductsContext";

//ContextProps
import {
  CategoriesContextProps,
  ProductsContextProps,
  UserContextProps,
  ReportsContextProps,
} from "./mockedContexts";

//Queries
import { LOGIN } from "../graphql/queries";

const mocksQuery: any = [
  {
    request: {
      query: LOGIN,
      variables: {
        email: "rick@email.com",
        password: "987654",
      },
    },
    result: {
      data: {
        login: {
          token: "1111",
          user: {
            id: "01",
            email: "myemail@email.com",
          },
          error: null,
        },
      },
    },
  },
];

const AllTheProviders: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  return (
    <RouterContext.Provider value={mockedRouter({})}>
      <UserContext.Provider value={UserContextProps}>
        <ReportsContext.Provider value={ReportsContextProps}>
          <CategoriesContext.Provider value={CategoriesContextProps}>
            <ProductsContext.Provider value={ProductsContextProps}>
              <MockedProvider mocks={mocksQuery} addTypename={false}>
                {children}
              </MockedProvider>
            </ProductsContext.Provider>
          </CategoriesContext.Provider>
        </ReportsContext.Provider>
      </UserContext.Provider>
    </RouterContext.Provider>
  );
};

const CustomRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export default CustomRender;
