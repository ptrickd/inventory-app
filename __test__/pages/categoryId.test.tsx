//Testing library
import { screen, cleanup, waitFor } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import userEvent from "@testing-library/user-event";

//Mocked router
const useRouter = jest.spyOn(require("next/router"), "useRouter");

// CustomRender
import CustomRender from "../functions/CustomRender";

//Page to test
import CategoryId from "../../pages/category/[categoryId]";

//Mocked Query
import { LOGIN } from "../../graphql/queries";

const mocksQuery: any = [
  {
    request: {
      query: LOGIN,
      variables: {
        email: "test@email.com",
        password: "securepassword11",
      },
    },
    result: {
      data: {
        login: {
          token: "1111",
          user: {
            id: "01",
            email: "test@email.com",
          },
          error: null,
        },
      },
    },
  },
  {
    request: {
      query: LOGIN,
      variables: {
        email: "test@email.com",
        password: "securepassword11",
      },
    },
    result: {
      data: {
        login: {
          token: "1111",
          user: {
            id: "01",
            email: "test@email.com",
          },
          error: null,
        },
      },
    },
  },
];

describe("<CategoryId />", () => {
  afterEach(() => {
    cleanup();
  });

  it("render as intended", async () => {
    CustomRender(
      <MockedProvider mocks={mocksQuery} addTypename={false}>
        <CategoryId />
      </MockedProvider>
    );

    //The page has the name of the category as title
    const categoryName = screen.getByText("Produce");
    expect(categoryName).toBeInTheDocument();

    //The page display a produce
    const productName = screen.getByText("Carrots");
    expect(productName).toBeInTheDocument();

    //The page show the current unit
    const unit = screen.getByText("ea");
    expect(unit).toBeInTheDocument();

    // The page has a button 'Add new product'
    const AddButton = screen.getByText("Add New Product");
    expect(AddButton).toBeInTheDocument();

    //the page has a 'next category button'
    const NextCategoryButton = screen.getByRole("button", {
      name: /next category/i,
    });
    expect(NextCategoryButton).toBeInTheDocument();

    //the page has a setting icon
    const SettingsIcon = screen.getByTestId(/settingsicon/i);
    expect(SettingsIcon).toBeInTheDocument();

    //has edit icon
    const EditIcon = screen.getByTestId(/editicon/i);
    expect(EditIcon).toBeInTheDocument();

    //has delete icon
    const DeleteIcon = screen.getByTestId(/deleteicon/i);
    expect(DeleteIcon).toBeInTheDocument();
  });

  it("open the edit modal when clicking on the edit icon", async () => {
    //mock router
    const router = { push: jest.fn(), query: { categoryId: "123456" } };
    useRouter.mockReturnValue(router);

    CustomRender(
      <MockedProvider mocks={mocksQuery} addTypename={false}>
        <CategoryId />
      </MockedProvider>
    );

    const user = userEvent.setup();

    //click on the edit icon
    const EditIcon = screen.getByTestId(/editicon/i);
    await user.click(EditIcon);

    //select the edit button of the modal to confirm ot opened
    const EditButton = screen.getByRole("button", { name: /edit/i });
    expect(EditButton).toBeInTheDocument();
  });

  it("delete the category when clicking on the delete icon", async () => {
    //mock router
    const router = { push: jest.fn(), query: { categoryId: "123456" } };
    useRouter.mockReturnValue(router);

    CustomRender(
      <MockedProvider mocks={mocksQuery} addTypename={false}>
        <CategoryId />
      </MockedProvider>
    );

    const user = userEvent.setup();

    //click on the delete icon
    const DeleteIcon = screen.getByTestId(/deleteicon/i);
    await user.click(DeleteIcon);
  });

  it("goes to the next category when clicking on the button", async () => {
    //mock router
    const router = { push: jest.fn(), query: { categoryId: "123456" } };
    useRouter.mockReturnValue(router);

    CustomRender(
      <MockedProvider mocks={mocksQuery} addTypename={false}>
        <CategoryId />
      </MockedProvider>
    );

    const user = userEvent.setup();

    //click on the next category button
    const NextCategoryButton = screen.getByRole("button", {
      name: /next category/i,
    });
    await user.click(NextCategoryButton);

    await waitFor(async () => {
      expect(router.push).toHaveBeenCalledTimes(1);
      expect(router.push).toHaveBeenCalledWith("123457");
    });
  });
  it("match the snapshot", () => {
    const tree = CustomRender(
      <MockedProvider mocks={mocksQuery} addTypename={false}>
        <CategoryId />
      </MockedProvider>
    );
    expect(tree).toMatchSnapshot();
  });
});
