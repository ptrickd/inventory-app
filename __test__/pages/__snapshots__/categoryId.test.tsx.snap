// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryId /> match the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    style=""
  >
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthXs ProductsPage-root css-1gbg9pz-MuiContainer-root"
      >
        <div
          class="ProductsPage-title"
        >
          <h2
            class="MuiTypography-root MuiTypography-h2 ProductsPage-title-text css-o2za6n-MuiTypography-root"
          >
            Produce
          </h2>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorIcon MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="EditIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
              />
            </svg>
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorIcon MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="DeleteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
              />
            </svg>
          </button>
        </div>
        <div>
          <h6
            class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
          />
        </div>
        <hr />
        <div
          class="InputProduct-root css-1rznzry"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
          >
            Carrots
          </h6>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth InputProduct-form-control css-ytlejw-MuiFormControl-root"
          >
            <form>
              <span
                class="InputProduct-inner-form-control"
              >
                <div
                  class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root InputProduct-textfield css-cmpglg-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-1ew92b2-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="true"
                    for="Carrotscurrent"
                    id="Carrotscurrent-label"
                  >
                    Current
                  </label>
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-6h9cc6-MuiInputBase-root-MuiInput-root"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiInput-input css-1yrc8ca-MuiInputBase-input-MuiInput-input"
                      id="Carrotscurrent"
                      name="currentAmount"
                      type="text"
                      value="6"
                    />
                  </div>
                </div>
              </span>
              <span
                class="InputProduct-inner-form-control"
              >
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl InputProduct-select-unit css-j218zi-MuiInputBase-root-MuiInput-root-MuiSelect-root"
                  data-testid="button"
                >
                  <div
                    aria-controls=":r1a:"
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    aria-labelledby="CarrotslabelID-select Carrotsselect"
                    class="MuiSelect-select MuiSelect-standard MuiInputBase-input MuiInput-input css-5lvf42-MuiSelect-select-MuiInputBase-input-MuiInput-input"
                    id="Carrotsselect"
                    role="combobox"
                    tabindex="0"
                  >
                    ea
                  </div>
                  <input
                    aria-hidden="true"
                    aria-invalid="false"
                    class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
                    name="selectUnit"
                    tabindex="-1"
                    value="ea"
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-1gpuzre-MuiSvgIcon-root-MuiSelect-icon"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
              </span>
            </form>
            <div
              class="InputProduct-box MuiBox-root css-w56oqh"
            >
              <span
                class="MuiTypography-root MuiTypography-caption InputProduct-last-amount-name css-du86ev-MuiTypography-root"
              >
                Last
              </span>
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-alignLeft InputProduct-last-amount-value css-1j3qii7-MuiTypography-root"
                style="--Typography-textAlign: left;"
              >
                5
              </p>
            </div>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorIcon MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="SettingsIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6"
                />
              </svg>
            </button>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary ProductsPage-button css-1abvrni-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Next Category
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary ProductsPage-button css-1abvrni-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Add New Product
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthXs ProductsPage-root css-1gbg9pz-MuiContainer-root"
    >
      <div
        class="ProductsPage-title"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2 ProductsPage-title-text css-o2za6n-MuiTypography-root"
        >
          Produce
        </h2>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorIcon MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="EditIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
            />
          </svg>
        </button>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorIcon MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="DeleteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
            />
          </svg>
        </button>
      </div>
      <div>
        <h6
          class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
        />
      </div>
      <hr />
      <div
        class="InputProduct-root css-1rznzry"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
        >
          Carrots
        </h6>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth InputProduct-form-control css-ytlejw-MuiFormControl-root"
        >
          <form>
            <span
              class="InputProduct-inner-form-control"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root InputProduct-textfield css-cmpglg-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-1ew92b2-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="true"
                  for="Carrotscurrent"
                  id="Carrotscurrent-label"
                >
                  Current
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-6h9cc6-MuiInputBase-root-MuiInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input css-1yrc8ca-MuiInputBase-input-MuiInput-input"
                    id="Carrotscurrent"
                    name="currentAmount"
                    type="text"
                    value="6"
                  />
                </div>
              </div>
            </span>
            <span
              class="InputProduct-inner-form-control"
            >
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl InputProduct-select-unit css-j218zi-MuiInputBase-root-MuiInput-root-MuiSelect-root"
                data-testid="button"
              >
                <div
                  aria-controls=":r1a:"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="CarrotslabelID-select Carrotsselect"
                  class="MuiSelect-select MuiSelect-standard MuiInputBase-input MuiInput-input css-5lvf42-MuiSelect-select-MuiInputBase-input-MuiInput-input"
                  id="Carrotsselect"
                  role="combobox"
                  tabindex="0"
                >
                  ea
                </div>
                <input
                  aria-hidden="true"
                  aria-invalid="false"
                  class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
                  name="selectUnit"
                  tabindex="-1"
                  value="ea"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-1gpuzre-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </div>
            </span>
          </form>
          <div
            class="InputProduct-box MuiBox-root css-w56oqh"
          >
            <span
              class="MuiTypography-root MuiTypography-caption InputProduct-last-amount-name css-du86ev-MuiTypography-root"
            >
              Last
            </span>
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-alignLeft InputProduct-last-amount-value css-1j3qii7-MuiTypography-root"
              style="--Typography-textAlign: left;"
            >
              5
            </p>
          </div>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorIcon MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="SettingsIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6"
              />
            </svg>
          </button>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary ProductsPage-button css-1abvrni-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Next Category
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary ProductsPage-button css-1abvrni-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Add New Product
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
